tools/apidocgen.py

```diff
@@ -0,0 +1,258 @@
+"""
+
+import sys
+import textwrap
+from importlib import import_module
+
+sys.path.insert(0, ".")
+
+from uaclient import data_types  # noqa: E402
+from uaclient.api.api import VALID_ENDPOINTS  # noqa: E402
+from uaclient.exceptions import UbuntuProError  # noqa: E402
+
+BLOCK_INDENT = "   "
+LIST_INDENT = "  "
+NO_ARGS = "- This endpoint takes no arguments."
+NO_FIELDS = "- This object has no fields."
+
+ENDPOINT_TEMPLATE = """\
+{name}
+{underline}
+
+{description}
+
+- Introduced in Ubuntu Pro Client Version: ``{introduced_in}~``
+
+Arguments:
+
+{args_section}
+
+.. tab-set::
+
+   .. tab-item:: Python API interaction
+      :sync: python
+
+      **Calling from Python code:**
+
+      .. code-block:: python
+
+         {example_python}
+
+      **Expected return object:**
+
+      {result_classes}
+
+      **Raised exceptions:**
+
+      {exceptions}
+
+   .. tab-item:: CLI interaction
+      :sync: CLI
+
+      **Calling from the CLI:**
+
+      .. code-block:: bash
+
+         {example_cli}
+
+      {example_cli_extra}
+
+      **Expected attributes in JSON structure:**
+
+      .. code-block:: js
+
+         {example_json}
+
+{extra}
+
+"""
+
+
+def data_value_type_to_str(data_value_cls):
+    if data_value_cls.__name__ == "_DataList":
+        return "List[{}]".format(
+            data_value_type_to_str(data_value_cls.item_cls)
+        )
+    if issubclass(data_value_cls, data_types.DataObject):
+        return data_value_cls.__name__
+    return data_value_cls.python_type_name
+
+
+def create_fields_table(cls):
+    table_template = """
+.. list-table::
+   :header-rows: 1
+
+   * - Field Name
+     - Type
+     - Description
+{fields}"""
+    field_definition_template = """
+* - ``{name}``
+  - ``{type}``
+  - {description}"""
+
+    if len(cls.fields) == 0:
+        return NO_FIELDS
+
+    fields = []
+    for field in cls.fields:
+        type_str = data_value_type_to_str(field.data_cls)
+        if not field.required:
+            type_str = "Optional[{}]".format(type_str)
+        fields.append(
+            textwrap.indent(
+                field_definition_template.format(
+                    name=field.key, type=type_str, description=field.doc or ""
+                ).strip(),
+                prefix=BLOCK_INDENT,
+            )
+        )
+    return table_template.format(
+        fields="\n".join(fields),
+    ).strip()
+
+
+def collect_data_object_classes(target_cls, collection, *, ignore=[]):
+    for field in target_cls.fields:
+        cls = field.data_cls
+        if cls.__name__ == "_DataList":
+            cls = cls.item_cls
+        if (
+            issubclass(cls, data_types.DataObject)
+            and cls not in collection
+            and cls not in ignore
+        ):
+            collection.append(cls)
+            collect_data_object_classes(cls, collection, ignore=[])
+
+
+def print_endpoint_docs(endpoint_name):
+    module = import_module("uaclient.api." + endpoint_name)
```

Comment:
I think we can have `try..except`  block here just to better inform when we provide an invalid endpoint

Reply:
Added!